Update the build and run scripts for NCAR machines

sjsprecious opened the pull request:
This PR updates the build and run scripts for NCAR machines (Derecho and Casper). Specifically:

- To build and run a case on Casper, just enter the `bin` folder and do `qsub casper_job_submit.sh`.
- To build and run a case on Derecho, just enter the `bin` folder and do `qsub derecho_job_submit.sh`.

To switch the CPU/GPU configuration or change the XML file for a different case, revise `ncar_build.sh` and `ncar_run.sh` scripts accordingly.

The new script will also generate a `run` folder under the root directory and run a case there. After the simulation is done, it will generate another folder under the `run` folder, whose name is case-dependent, and move all the output to the newly generated folder. In this way, the root directory won't get messed up anymore.

johnmauff commented on bin/casper_job_submit.sh:
```diff
@@ -0,0 +1,36 @@
+# Build the code #
+##################
+cd ..
+./bin/ncar_build.sh gpu
```
@sjsprecious I think we need more flexibility here.  We want to be able to run both the CPU and GPU versions of the code easily.  Just hard coding this for GPU execution is not what we want.

sjsprecious replied:
Thanks @johnmauff . A user can remove the `gpu` option in this script to build a CPU case instead. I just use a GPU case as an example. Note that this is a PBS script so I don't think it will take any argument to switch between CPU and GPU easily. Correct me if I am wrong here.

johnmauff replied:
Maybe a workaround would be to have GPU and CPU versions of the run scripts.  When we set up our CI/CD we will want the ability easily check on both CPU and GPU platforms

sjsprecious replied:
Thanks @johnmauff for the suggestion. That will work. Do you want me to generate separate run scripts for V100, A100, etc as well? Or we should ask the user to change the architecture based on their desires?

johnmauff replied:
I think, for now, just a single GPU script is fine.  We may need to do something a bit more sophisticated depending on the platforms in which we apply CI/CD.

johnmauff commented on bin/ncar_run.sh:
```diff
@@ -1,7 +1,8 @@
 
-# ./build/release/bin/samurai does not work somehow
-EXE="./build/release/bin/samurai"
+# This script should be executed under /path_to_samurai_root/run directory
```
@sjsprecious we need to have the ability to run this script on the CPU using OpenMP threads.  Currently, this script only runs a single GPU or CPU.  If on the CPU we should set the number of OpenMP threads in this file.

sjsprecious replied:
My bad! I thought we did not want to deal with the OpenMP threading so I just deleted those settings. Add them back now.

johnmauff commented on bin/ncar_build.sh:
```diff
@@ -0,0 +1,71 @@
+  # Somehow the Release mode will break the GPU code; change back to Debug mode
+  sed -i 's/CMAKE_BUILD_TYPE Release/CMAKE_BUILD_TYPE Debug/g' CMakeLists.txt
+  if [ "$NCAR_HOST" == "casper" ]; then
+    # Use cc70 for V100 GPU on Casper
```
@sjsprecious Casper has both V100 and A100 nodes.  I believe that H100 nodes will be installed on Casper sometime in the summer.    We can't assume that we will only be using the V100 nodes on Casper

sjsprecious replied:
Good point. I have removed this hardcode change so a user needs to revise the CMakeLists file manually for the desired GPU architecture.

johnmauff commented on bin/ncar_build.sh:
```diff
@@ -0,0 +1,71 @@
+fi
+
+# If we are going to measure the performance for Supercell and Hurricane,
+# we only want to run the first 20 iterations for each case.
```
@sjsprecious the paper has results for both 20 iterations and the full number.  It is not a good idea to force the change to 20 iterations here. If we only look at 20 iterations, we will likely miss numerical problems that cause the code to fail to converge in the expected number of iterations.

sjsprecious replied:
Good point. I have removed this hardcode change.

johnmauff reviewed (changes requested):
Thanks for the nice additions to save specific runs to unique directories.  I think we still need multiple changes to increase the flexibility of the run scripts.

mmbell commented:
The capability to read and write from other directories is already present in the parameter specification. "data_directory" can provide an absolute path to the input location, and "output_directory" provides a path to the desired output location. I believe it will fail though if either of these two directories are missing rather than creating them. There is currently no capability to create a root/run directory that lives above the data/output, as this has typically been done manually by the user. If there is desire for a "run" directory from the script for use on NCAR machines that is fine, it just needs to then be consistent with data/output parameters or that run directory will probably be ignored.

sjsprecious commented:
> The capability to read and write from other directories is already present in the parameter specification. "data_directory" can provide an absolute path to the input location, and "output_directory" provides a path to the desired output location. I believe it will fail though if either of these two directories are missing rather than creating them. There is currently no capability to create a root/run directory that lives above the data/output, as this has typically been done manually by the user. If there is desire for a "run" directory from the script for use on NCAR machines that is fine, it just needs to then be consistent with data/output parameters or that run directory will probably be ignored.

Thanks Michael for providing this information. I did not realize those two options before.

We typically run SAMURAI through `./samurai /path_to_xml_file`. Do you mean that we can also run something like `./samurai /path_to_xml_file --data_directory /path_to_data --output_directory /path_to_output`?

In addition, the XML file already contains the path to the input data. Thus what is `data_directory` used for in this case?

johnmauff commented on bin/ncar_run.sh:
```diff
@@ -0,0 +1,26 @@
+
+# This script should be executed under /path_to_samurai_root/run directory
+
+EXE="../build/release/bin/samurai"
```
This script still appears to make assumptions about where it is run.  I find this problematic.  We need a cleaner way that does not break when Samurai is not executed from the bin directory

sjsprecious replied:
So maybe you prefer an absolute rather than relative path here?

johnmauff reviewed (changes requested):
I still have concerns regarding the hardcoded paths to the executable used in the run scripts.  We really need a more flexible approach that does not break if the scripts are used outside of the bin directory.

sjsprecious commented:
> I still have concerns regarding the hardcoded paths to the executable used in the run scripts. We really need a more flexible approach that does not break if the scripts are used outside of the bin directory.

Thanks @johnmauff for your comment. I have updated the build and run scripts, and ask a user to explicitly set an environment variable named `SAMURAI_ROOT` so that we are able to run the NCAR scripts in a different path. I also moved all the NCAR scripts to a different folder named `ncar_scripts`.

johnmauff commented:
Thank you for these modifications.  These changes now look good.  I am currently running tests on a single node of Derecho for the Beltrami, supercell, and hurricane test configurations.
